### DISCUSSION TITLE
Split project into three

### DISCUSSION THREAD

**Discussion by sksamuel** [kotest/kotest#4056] [6.0 changes]
I've brought this up a few times, but generally my thinking is:

proposal:
kotest (main repo stays as the home of the docs)
kotest-framework (new repo)
kotest-property (new repo)
kotest-assertions (new repo)

[vert.x](https://github.com/eclipse-vertx) kind of do this pretty well 

pros:
- the project is huge, takes a while to compile / reindex etc. 
- Smaller repos are easier for people to contribute to.
- Three different scopes makes it easier for maintainers to specialize. I find it tricky keeping up with all the changes, so tend to just focus on the framework PRs.
- you don't need to worry about all the tricky junit stuff if you just wanna add a property test feature or an assertion.
- can release at different cadences, especially useful for keeping up with KMP changes

cons:
- people will invariably not read the changelog and be confused about why the versions are now out of step
- harder to make breaking changes since different versions have to work together (although I think in practice this isn't so much of a big deal as we aren't making wholesale changes anymore)

**Comment by OliverO2** [comment]
I like the idea of having an independent framework repo. If "kotest-property" and "kotest-assertions" have some minimum requirement on the framework, _if_ they are used in conjunction with the framework, can we figure out a mechanism to detect this and inform users when the requirement is not met?

**Reply by sksamuel** [reply]
So far the only shared modules are the `kotest-common` which is mostly some helpers that are useful in KMP, and could just be copied into each repo, and the `kotest-assertions-shared` module which contains all the Eq and exceptions stuff which the framework depends on for generating nicely formatted errors. I think it would be trivial to break this apart, or worse case we can still have a kotest-core module that just contains helpers for all three projects.

I always had the mindset that property testing, assertions and the framework should be three individual things you can opt in to. Lots of people I imagine use kotest without kotest assertions (eg with power assert) and I imagine some people use kotest assertions in junit.

**Reply by OliverO2** [reply]
Yep. Decoupling where possible is always a good idea. We'd just have to consider cases like this one, where the assertions rely on proper setup as done by the framework:

https://github.com/kotest/kotest/blob/29ed6c9f7753256fcbdfd7e2fddc99b2ff2e4392/kotest-assertions/kotest-assertions-shared/src/jvmMain/kotlin/io/kotest/assertions/ErrorCollector.kt#L13-L20

**Comment by aSemy** [comment]
Something to consider is whether Kotest's Gradle build conventions should be shared between the split projects. Probably yes. In which case, the conventions would need to be published.

As an incremental step, you could split up the project as you described, but keep them in the same repo, and use [composite builds](https://docs.gradle.org/current/userguide/composite_builds.html).

**Reply by OliverO2** [reply]
Good point!

**Reply by Kantis** [reply]
I think getting the convention plugins published would be a good step either way. It's currently a bit of a mess to manage gradle configuration across all the extension repositories.

Maybe we could take some inspiration from https://github.com/arrow-kt/arrow-gradle-config

**Reply by sksamuel** [reply]
We could do that yeah, kotest-gradle-config

**Comment by aSemy** [comment]
> * the project is huge, takes a while to compile / reindex etc.

A quick win for improving Gradle configuration time would be to enable configuration cache. It's currently disabled, but since the project has updated to use KGP 1.9 it should be re-enabled.

https://github.com/kotest/kotest/blob/77cf524cb89dc04c8b7f075efa770c6249f020e1/gradle.properties#L28-L29

I'll make a PR... https://github.com/kotest/kotest/pull/4062

Another relatively quick win would be setting up a remote build cache server - although it would require a server!

**Comment by OliverO2** [comment]
What if we'd try the quick wins first and then see if they are just good enough for us?

1. #4062
2. Same repo, composite builds.

We'd not gain all the pros, but avoid the following cons for now:
* A higher volume of user questions regarding versions.
* Easier cross-project analysis and changes (find usages, search+replace).

After gaining some experience with the quick-win setup, we'd have better information for deciding whether to split into multiple repos later.

**Reply by aSemy** [reply]
Sounds good to me! 

Just to temper expectations, there are some drawbacks with composite builds. The most annoying one is that Gradle won't 'propagate' tasks from the parent build to children https://github.com/gradle/gradle/issues/20863, but there are some workarounds.

**Comment by christophsturm** [comment]
I like the idea. 
I also like @OliverO2 s idea of trying it out in one repo first, and then moving to separate repos. 
Also something to consider is maybe is there a better name for kotest-framework? (framework imo is like util, a name that means not much). also if you want to encourage usage of the assertion and property testing lib outside of kotest maybe the names should be more decoupled, so one example just to get the idea rolling could be

* kotest-framework => ko-test 
* kotest-assertion => ko-assert
* kotest-property => ko-property

or maybe kt-test , kt-assert, kt-property?

**Reply by OliverO2** [reply]
Wikipedia offers this definition:
> A framework is a generic term commonly referring to an essential supporting structure which other things are built on top of.

I'd say the term "kotest-framework" is pretty accurate. It defines the test structure and provides an engine operating on that structure. The name "ko-test" sounds less indicative to me. "kt" would be read as "Kotlin" and clash with "kotlin-test".

**Reply by sksamuel** [reply]
Or kotest-engine if we really want something other than framework.

**Reply by OliverO2** [reply]
Engine seems too narrow. There are those Specs which are not part of the engine.

But while we're at naming: I could never really figure out what the difference between a runner and an engine might be...

**Reply by sksamuel** [reply]
I think for 6.0 I am thinking we should collapse all the kotest-framework modules together.
Have `kotest-framework-engine` (or just kotest-framework) and then a junit runner as an optional extra, plus the gradle plugin bits.

**Reply by christophsturm** [reply]
> Wikipedia offers this definition:
> 
> > A framework is a generic term commonly referring to an essential supporting structure which other things are built on top of.
> 
> I'd say the term "kotest-framework" is pretty accurate. 

but if the 3 parts are to be independent of each other then its not accurate at all. "framework" is the test runner. 
ok

**Reply by Kantis** [reply]
There's extensions, tags, discovery, spec styles, etc.. Essentially everything to build and run tests, minus assertions. I think it's fair to call it a `framework`

**Reply by sksamuel** [reply]
If we don't like the names what about 

Kotest
Koassert
Kocheck

**Reply by sksamuel** [reply]
Might make the version reset clearer too

**Reply by Kantis** [reply]
I think that would be way more confusing. Kotest is a pretty established name already. We could just change the artifact coordinates and start 6.0 there?

Framework -> `io.kotest.framework:$module`
Assertions -> `io.kotest.assertions:$module`
etc...

**Comment by AlexCue987** [comment]
should we start small: sprout out just one thing first and see how it works out? maybe move out `kotest-assertions` as the first step?

**Comment by sksamuel** [comment]
Personally I'm not really in favor of composite builds because the whole repo is already very complex to newcomers.
Try cloning the intellij-community repo. It takes a month just to figure out how to build it because of all the plugins in there.

This would make the gradle build setup even more complex to handle running multiple projects in the same repo.

In general I prefer a little bit of copy n paste for things that are very rarely changing (like our build scripts).

**Comment by sksamuel** [comment]
Decided against this
